### PR TITLE
klocwork 2021.4

### PR DESCRIPTION
**Pull request by mandrono**
Issue: 57193

### CODE REVIEW

**Comment by mandrono** [comment]
@apankratovantonp @ilya-lavrenov there are some issues: "Class has virtual methods inherited from a base class, but its destructor is not virtual and not empty" for MKLDNNAsyncInferRequest, MKLDNNInferRequest, Engine, CNNNetworkImpl.
could you take a look, please?

@nshchego could you take a look at the rest, please?

**Comment by apankratovantonp** [comment]
@mandrono Please mark "Class has virtual methods inherited from a base class, but its destructor is not virtual and not empty" As `not a problem` and revert corresponding changes.

**Comment by mandrono** [comment]
> @mandrono Please mark "Class has virtual methods inherited from a base class, but its destructor is not virtual and not empty" As `not a problem` and revert corresponding changes.

reverted

**Review by apankratovantonp — APPROVED** [review]

**Review by ilya-lavrenov — APPROVED** [review]

File: docs/snippets/InferenceEngine_network_with_state_infer.cpp [hunk 1]
```diff
@@ -64,7 +64,12 @@ int main(int argc, char *argv[]) {
             inferRequest.Infer();
             // check states
             auto states = inferRequest.QueryState();
+            if (states.empty())
+                continue;
```
Comment (nshchego) [inline docs/snippets/InferenceEngine_network_with_state_infer.cpp:68]: Perhaps exception is needed?

Comment (mandrono) [reply docs/snippets/InferenceEngine_network_with_state_infer.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.h [hunk 2]
```diff
@@ -32,7 +32,7 @@ class MKLDNNGatherElementsNode : public MKLDNNNode {
     size_t dataTypeSize_;
     int strideAxDst_;
     int dstAxDim_;
-    int strideAx1Diff_;
+    int strideAx1Diff_ = 0;
```
[the remark this answers is not in the document]
Comment (mandrono) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_elements_node.h:35]: these variable initialize in ctor, i guess we don't need to initialized them in header

**Review by ilyachur — APPROVED** [review]

**Review by nshchego — APPROVED** [review]

**Pull request merged**